### PR TITLE
Ipynb examples

### LINKED ISSUE #335: iPython notebooks for existing examples
It would be good to have iPython notebooks showing the results of the examples to guide the user through.

Something like this (for the MCCL galaxies):
https://github.com/skypyproject/skypy/pull/284

This could also be linked to a binder (a la https://mybinder.org/v2/gh/ntessore/skypy-0.3-demo/master) allowing the user to interact with the notebooks online as well.

This could also be expanded in future to new examples, other demos etc.

### PR DESCRIPTION
**Pull request by itrharrison**
## Description
Merging will close #335. Have added iPython notebooks making plots from the outputs of the current example config files.

## Checklist
- [X] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [ ] Write unit tests
- [ ] Write documentation strings
- [ ] Assign someone from your working team to review this pull request
- [X] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

File: skypy/power_spectrum/tests/test_camb.py [hunk 1]
```diff
@@ -35,4 +35,4 @@ def test_camb():
     redshift = [1.0, 0.0]
     pzk = camb(wavenumber, redshift, Planck15, 2.e-9, 0.965)
     assert pzk.shape == (len(redshift), len(wavenumber))
-    assert allclose(pzk, test_pzk[np.argsort(redshift), :], rtol=1.e-4)
+    assert allclose(pzk, test_pzk[np.argsort(redshift), :], rtol=1.e-4)
```
Comment (Lucia-Fonseca) [inline skypy/power_spectrum/tests/test_camb.py:38]: it complains "no newline at the end of file"

File: skypy/power_spectrum/tests/test_class.py [hunk 2]
```diff
@@ -44,3 +44,4 @@ def test_classy():
     wavenumber = 1.e-1
     pzk = classy(wavenumber, redshift, Pl15massless)
     assert np.isscalar(pzk)
+
```
Comment (Lucia-Fonseca) [inline skypy/power_spectrum/tests/test_class.py:47]: it complains "no newline at the end of file"

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]

File: examples/abundance_matching.ipynb [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(halos[halos['parent_halo']]['halo_mass'], histtype='step', bins=np.logspace(9,12,50), label='Parent halos')\n",
+    "plt.hist(halos[~halos['parent_halo']]['halo_mass'], histtype='step', bins=np.logspace(9,12,50), label='Sub halos')\n",
+    "plt.xlabel('$M\\,[M_{\\odot}]$')\n",
+    "plt.xscale('log')\n",
+    "plt.yscale('log')\n",
+    "plt.legend()"
```
Comment (Lucia-Fonseca) [inline examples/abundance_matching.ipynb:131]:
```suggestion
    "plt.legend()"
    "plt.show()"
```

File: examples/abundance_matching.ipynb [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(halos['galaxy_magnitude'], histtype='step', bins=50)\n",
+    "plt.xlabel('$m$')\n",
+    "plt.yscale('log')"
```
Comment (Lucia-Fonseca) [inline examples/abundance_matching.ipynb:171]:
```suggestion
    "plt.yscale('log')"
    "plt.show()"
```

File: examples/abundance_matching.ipynb [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(halos['halo_group'], histtype='step', bins=halos['halo_group'].max())\n",
+    "plt.xlabel('Group')"
```
Comment (Lucia-Fonseca) [inline examples/abundance_matching.ipynb:212]:
```suggestion
    "plt.xlabel('Group')"
    "plt.show()"
```

File: examples/abundance_matching.ipynb [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist2d(halos['halo_mass'], halos['galaxy_magnitude'], bins=[np.logspace(9,12,50), 50])\n",
+    "plt.xscale('log')\n",
+    "plt.xlabel('$M\\,[M_{\\odot}]$')\n",
+    "plt.ylabel('$m$')"
```
Comment (Lucia-Fonseca) [inline examples/abundance_matching.ipynb:255]:
```suggestion
    "plt.ylabel('$m$')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist2d(merger_rates['redshift'], merger_rates['magnitude'], bins=50)\n",
+    "plt.xlabel('$z$')\n",
+    "plt.ylabel('$m$')"
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:134]:
```suggestion
    "plt.ylabel('$m$')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(merger_rates['redshift'], histtype='step', bins=50)\n",
+    "plt.xlabel('$z$')"
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:168]:
```suggestion
    "plt.xlabel('$z$')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(merger_rates['magnitude'], histtype='step', bins=50)\n",
+    "plt.xlabel('$m$')"
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:202]:
```suggestion
    "plt.xlabel('$m$')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(merger_rates['luminosity'], histtype='step', bins=np.logspace(5,12,50))\n",
+    "plt.xlabel('$\\phi\\,$[$L_{\\odot, B}$]')\n",
+    "plt.xscale('log')\n",
+    "plt.yscale('log')"
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:240]:
```suggestion
    "plt.yscale('log')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(merger_rates['ns_ns_rate'], histtype='step', bins=np.logspace(-4,2,50), label='NS-NS')\n",
+    "plt.hist(merger_rates['ns_bh_rate'], histtype='step', bins=np.logspace(-4,2,50), label='NS-BH')\n",
+    "plt.hist(merger_rates['bh_bh_rate'], histtype='step', bins=np.logspace(-4,2,50), label='BH-BH')\n",
+    "plt.legend()\n",
+    "plt.xscale('log')\n",
+    "plt.yscale('log')\n",
+    "plt.xlabel('$R\\,$[yr$^{-1}$]')"
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:300]:
```suggestion
    "plt.xlabel('$R\\,$[yr$^{-1}$]')"
    "plt.show()"
```

File: examples/gravitational_wave_rates.ipynb [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This example runs a number of SkyPy modules to create a catalogue of galaxies with compact binary merger rates according to the linear model of [Abadie et al (2010)](ttps://arxiv.org/abs/1003.2480) (Table III).\n",
+    "\n",
+    "Running the `merger_rates` example:\n",
+    "```bash\n",
+    "$ skypy merger_rates.yml --format fits\n",
```
Comment (Lucia-Fonseca) [inline examples/gravitational_wave_rates.ipynb:34]: Is the name of the file `merger_rates.yml` or `gravitational_wave_rates.yml`?

File: examples/mccl_galaxies.ipynb [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist2d(red_galaxies['redshift'], red_galaxies['magnitude'], bins=50)\n",
+    "plt.xlabel('$z$')\n",
+    "plt.ylabel('$m$')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:129]:
```suggestion
    "plt.ylabel('$m$')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(red_galaxies['redshift'], histtype='step', bins=50)\n",
+    "plt.xlabel('$z$')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:163]:
```suggestion
    "plt.xlabel('$z$')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(red_galaxies['magnitude'], histtype='step', bins=50)\n",
+    "plt.xlabel('$m$')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:197]:
```suggestion
    "plt.xlabel('$m$')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(red_galaxies['physical_size'], histtype='step', bins=50)\n",
+    "plt.xlabel('$d\\,$[kpc]')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:247]:
```suggestion
    "plt.xlabel('$d\\,$[kpc]')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(red_galaxies['ellipticity'], histtype='step', bins=50)\n",
+    "plt.xlabel('$|e|$')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:294]:
```suggestion
    "plt.xlabel('$|e|$')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "plt.hist(red_galaxies['physical_size'], histtype='step', bins=50)\n",
+    "plt.xlabel('$\\\\theta\\,$[rad]')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:340]:
```suggestion
    "plt.xlabel('$\\\\theta\\,$[rad]')"
    "plt.show()"
```

File: examples/mccl_galaxies.ipynb [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.figure()\n",
+    "for spc in red_galaxies['spectral_coefficients'].T:\n",
+    "    plt.hist(spc, histtype='step', bins=50)\n",
+    "plt.yscale('log')\n",
+    "plt.xlabel('$c_{i}$')"
```
Comment (Lucia-Fonseca) [inline examples/mccl_galaxies.ipynb:390]:
```suggestion
    "plt.xlabel('$c_{i}$')"
    "plt.show()"
```

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]
It looks very good! Great job, Ian.
Two small changes:
* Add `plt.show()` to avoid outputing unnecesary text
* Name of the config file for merger rates
* I still see `test_camb.py` and `test_class.py`.

**Comment by jucordero** [comment]
I think you can also do
`plt.xlabel('This is an x label');`
to avoid extra text output in the notebooks without the need of a `plt.show`

**Comment by Lucia-Fonseca** [comment]
That works too, thanks @jucordero!
> I think you can add a semicolon at the end of a statement
> > `plt.xlabel('This is an x label');`
> > to avoid extra text output in the notebooks without the need of a `plt.show`

File: examples/gravitational_wave_rates.ipynb (hunk 12 above)
Comment (Lucia-Fonseca) [reply examples/gravitational_wave_rates.ipynb:34]: This still needs to be addressed.

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]

**Review by rrjbca — COMMENTED** [review]
Is the intention that the user would run these notebooks themselves? I think they make more sense as part of the documentation (perhaps using [nbsphinx](https://nbsphinx.readthedocs.io/en/0.7.1/)?) but if that is the case they should probably be in the `docs` directory not the `examples` directory. Particularly because everything in `examples` is mirrored to the [`skypyproject/examples`](https://github.com/skypyproject/examples) repository and we shouldn't distribute them in this way if we don't expect the user to run them.

**Comment by Lucia-Fonseca** [comment]
The idea is to have something similar to CAMB DEMO 
https://camb.readthedocs.io/en/latest/CAMBdemo.html

**Review by Lucia-Fonseca — APPROVED** [review]
Looks good.

File: skypy/power_spectrum/tests/test_camb.py (hunk 1 above)
Comment (rrjbca) [inline skypy/power_spectrum/tests/test_camb.py:38]:
```suggestion
    assert allclose(pzk, test_pzk[np.argsort(redshift), :], rtol=1.e-4)

```

File: skypy/power_spectrum/tests/test_class.py (hunk 2 above)
Comment (rrjbca) [inline skypy/power_spectrum/tests/test_class.py:47]:
```suggestion

```

**Review by rrjbca — APPROVED** [review]
These look good, let's merge them as they are and work out how we want to incorporate them with the documentation in the future.

**Pull request merged**